test/test_autograd.py

```diff
@@ -3204,6 +3204,361 @@ def backward(ctx, gO):
                     out.backward()
             self.assertIn('MyFunc.apply', str(w[0].message))
 
+    @unittest.skipIf(not torch.cuda.is_available(), "CUDA unavailable")
+    @skipIfRocm
+    def test_grad_scaling_builtins(self):
+        current_scale = torch.cuda.FloatTensor([4.0])
+
+        found_inf = torch.cuda.FloatTensor([0.0])
+        g = torch.cuda.FloatTensor(1000).fill_(4.0)
+        torch._amp_unscale_inf_check_(g, current_scale, found_inf)
+        self.assertTrue(found_inf.item() == 0.0)
+        self.assertTrue(torch.allclose(g, torch.ones(1000, device="cuda", dtype=torch.float32)))
+
+        found_inf.zero_()
+        g = torch.cuda.FloatTensor([float('inf')])
+        torch._amp_unscale_inf_check_(g, current_scale, found_inf)
+        self.assertTrue(found_inf.item() == 1.0)
+
+        found_inf.zero_()
+        g = torch.cuda.FloatTensor([float('nan')])
+        torch._amp_unscale_inf_check_(g, current_scale, found_inf)
+        self.assertTrue(found_inf.item() == 1.0)
+
+        growth_factor = 4.0
+        backoff_factor = 0.5
+
+        found_inf.zero_()
+        new_scale = torch._amp_update_scale(current_scale, found_inf, growth_factor, backoff_factor)
+        self.assertTrue(new_scale.item(), 16.0)
+
+        found_inf.fill_(1.0)
+        new_scale = torch._amp_update_scale(current_scale, found_inf, growth_factor, backoff_factor)
+        self.assertTrue(new_scale.item(), 2.0)
+
+    def _create_scaling_models_optimizers_S(self):
+        mod_control = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        mod_scaling = torch.nn.Sequential(torch.nn.Linear(8, 8), torch.nn.Linear(8, 8)).cuda()
+        # model_scaling.load_state_dict(model_control.state_dict()) might also work
+        for c, s in zip(mod_control.parameters(), mod_scaling.parameters()):
+            s.data.copy_(c.data)
+
+        opt_control = torch.optim.SGD(mod_control.parameters(), lr=1.0)
+        opt_scaling = torch.optim.SGD(mod_scaling.parameters(), lr=1.0)
+
+        S = torch.amp.add_amp_attributes(opt_scaling)
+
+        return mod_control, mod_scaling, opt_control, opt_scaling, S
+
+    def _create_scaling_case(self):
+        mod_control, mod_scaling, opt_control, opt_scaling, S = self._create_scaling_models_optimizers_S()
+
+        # There's no need to use 2**24 for functional correctness testing.
```

Comment:
What is this comment referring to?

Reply:
The default initial scale value (S given by `S = torch.amp.add_amp_attributes(optimizer)`) is 2**24.

Reply:
Let's try to clarify the comment.